### ISSUE TITLE
Cropping polygons are sometimes blurry near edge

### ISSUE THREAD

**Issue by maxbogue** [openseadragon/openseadragon#2312]
Hello! I'm using TiledImage.setCroppingPolygons to crop an overlay that appears on top of a main image, and I'm seeing some weird behavior. As you zoom, the edge sometimes gets blurry and pulls away from the polygon edge a little bit (maybe 10-20px of the main image), but then will snap back as you zoom in. If the viewer is rotated at all, the cropped overlay can jump several hundred pixels, typically up and to the left relative to the unrotated main image.

These seem very related but are quite confusing; if it was just the offset I'd think we were computing the overlay position of the polygons incorrectly but the blurred edge issue (which seems to happen at the same zoom level but rotation 0) is a whole different story. Based on [this condition and TODO](https://github.com/openseadragon/openseadragon/blob/ead728dc3519bc578a0e29642445a6ca8837431f/src/tiledimage.js#L1875) I'm guessing the issue has something to do with the "sketch canvas".

Any help I could get with this issue would be greatly appreciated!

**Comment by pearcetm** [comment]
Any chance you can link to an example (codepen or similar) that reproduces the issue?

**Comment by maxbogue** [comment]
Difficult to get an env set up but here's a quick video of it:

https://user-images.githubusercontent.com/299163/223173234-7debf539-ba6c-4129-8e4e-2a298495ad53.mov

**Comment by maxbogue** [comment]
And here is it jumping when the image is rotated:

https://user-images.githubusercontent.com/299163/223174271-d9743f2d-0209-4b76-a29d-85496ab4ce5c.mov

**Comment by pearcetm** [comment]
Thanks for the videos. Can you provide some more detail about the overlay layer(s) - both the red/transparent layer, and the polygon? Are you using one of the overlay plugins?

**Comment by maxbogue** [comment]
The polygon is actually being drawn by a different library on a different canvas (something we'd like to fix in the future but for now it shouldn't be relevant). The overlays are just their own DZIs being added as tiled images with `opacity` and `fitBounds` set.

**Comment by maxbogue** [comment]
I'm happy to help with a fix here if someone can give me some guidance on what's going on! I sent a message in the Discord #help channel as well if it's easier to discuss there.

**Comment by maxbogue** [comment]
If it's any help, commenting out this entire block of code seems to fix the problem:

![Screenshot 2023-03-06 at 2 08 52 PM](https://user-images.githubusercontent.com/299163/223206990-df99c89d-212b-4bc5-97d2-3d2bad13005f.png)

Though obviously that's probably breaking something else less obvious to me. Pretty difficult to isolate where things are going wrong because nearly every branch of that function references `sketchScale` and `sketchTranslate`.

**Comment by pearcetm** [comment]
You can try setting `smoothTileEdgesMinZoom` to `Infinity` to disable that block. Does it generate seams in your tiled image(s) if you do so?

**Comment by maxbogue** [comment]
Oh that's quite convenient. I'm unable to see any seams; in what scenario would they be visible?

**Comment by pearcetm** [comment]
I'm not entirely sure. It seems from the comment in the code you included that, for some tile sources at least, high zoom levels can lead to visible seams between tiles.

Coincidentally, I've been working on a PR (https://github.com/openseadragon/openseadragon/pull/2310) that addresses a lot of the problems related to visual tiling artifacts by moving towards `WebGL` based rendering. You may find that to be of interest!

**Comment by iangilman** [comment]
`smoothTileEdgesMinZoom` was introduced with https://github.com/openseadragon/openseadragon/pull/764 which fixed https://github.com/openseadragon/openseadragon/issues/755. It only affects when you are zoomed in more than full resolution, and you only really see the effects when you're zoomed in a lot past full resolution (which generally is never the case). 

I think disabling it in this scenario is a fine workaround, but of course hopefully someone can look into fixing this bug as well!

**Comment by pearcetm** [comment]
To fix the blurring/jitter during zoom, hard-coding `sketchScale=1` instead of the computed value `` does the trick:

```
        if (lastDrawn.length > 1 &&
            imageZoom > tiledImage.smoothTileEdgesMinZoom &&
            !tiledImage.iOSDevice &&
            tiledImage.getRotation(true) % 360 === 0 && // TODO: support tile edge smoothing with tiled image rotation.
            $.supportsCanvas && this.viewer.useCanvas) {
            // When zoomed in a lot (>100%) the tile edges are visible.
            // So we have to composite them at ~100% and scale them up together.
            // Note: Disabled on iOS devices per default as it causes a native crash
            useSketch = true;
            // sketchScale = tile.getScaleForEdgeSmoothing();                                        // Instead of this
            sketchScale = 1;                                                                                                  // Just set the scale to 1
            sketchTranslate = tile.getTranslationForEdgeSmoothing(sketchScale,
                this.getCanvasSize(false),
                this.getCanvasSize(true));
        }
```


To fix the mis-positioning when rotated, adding the lines highlighted below works:

```
        if (tiledImage._croppingPolygons) {
            var self = this;
            this.saveContext(useSketch);
            try {
                var polygons = tiledImage._croppingPolygons.map(function (polygon) {
                    return polygon.map(function (coord) {
                        var point = tiledImage
                            .imageToViewportCoordinates(coord.x, coord.y, true)
                            .rotate(-tiledImage.getRotation(true), tiledImage._getRotationPoint(true));
                        var clipPoint = self.viewportCoordToDrawerCoord(point);
                        if (sketchScale) {
                            clipPoint = clipPoint.times(sketchScale);
                        }
                        if (sketchTranslate) {                                                   // NEW LINES
                            clipPoint = clipPoint.plus(sketchTranslate);         // NEW LINES
                        }                                                                                    // NEW LINES
                        return clipPoint;
                    });
                });
                this.clipWithPolygons(polygons, useSketch);
            } 

```

Note that this entire snippet is copied from my ongoing implementation that separates drawing from positioning etc, and the surrounding code may have been modified from current OSD core code (and thus to be applied to current OSD it would have to be modified as well).

Something I noticed while testing is that if `smoothTileEdgesMinZoom` is set to a value below 1 (0.1 demonstrates this nicely), the `sketchCanvas` path gets very messed up and truncates zoomed-out images, suggesting the entire code path either needs some guards or needs to be re-thought.

However, when fixing the issue with the changes above (which avoid calculating a scale factor for the sketchCanvas), even at extreme zoom levels I do not see tiling issues... has the fix that introduced this code been superseded by subsequent fixes? Or maybe by improvements in `canvas` implementations by supported browsers? Even without these changes, I haven't been able to reproduce the original issue with tiling even at extreme zoom levels, with the fix using `sketchCanvas` disabled. I am not sure what I should to expect to be broken by removing the calculation of `sketchScale` from the equation.

**Comment by iangilman** [comment]
@pearcetm Interesting! It certainly would be cool if we didn't need the edge smoothing aspect of sketchCanvas anymore. Note that we still need sketchCanvas for semi transparent images regardless.

As for reproducing the issue, the top comment in https://github.com/openseadragon/openseadragon/issues/755 has a screenshot at 800% that shows it. Maybe try something like that (if you want to verify the repro or lack thereof).

**Comment by iangilman** [comment]
@pearcetm I suppose the blurring us because the polygon is being applied in the sketch canvas, so when it's blown up to the main canvas, it becomes blurry? If that's the case, the solution would be to apply the polygon during the transfer to the main canvas. I'm not sure how involved that change would be.

Even without that fix, fixing all of the other buggy behavior seems worthwhile! Do you want to make a PR with that?

**Comment by pearcetm** [comment]
> the solution would be to apply the polygon during the transfer to the main canvas. I'm not sure how involved that change would be.

Yes, that's exactly the problem. The refactoring needed to address it is not entirely trivial, since tiles are composited onto the main canvas from the (scaled) sketch canvas individually at the moment. Applying cropping during transfer would seem to require adding an intermediate canvas scaled to match the main canvas, which the sketch would need to be drawn onto, cropped, and then drawn to the main canvas. This would double the number of drawing operations and require an extra canvas. Doable, though with a performance cost.

A PR that addresses at least the translation-based bug has been submitted: https://github.com/openseadragon/openseadragon/pull/2316

**Comment by iangilman** [comment]
#2316 fixed the translation issue but not the blur. I'll reopen this in case someone wants to tackle the blur. That said, it sounds like @maxbogue doesn't need it at the moment.
